src/canvas/getHoveredEvent.js

```diff
@@ -0,0 +1,241 @@
+ * currently being hovered over.
+ */
+function getHoveredFlamechartEvent(
+  data: $ReadOnly<ReactProfilerDataV2>,
+  flamechart: $ReadOnly<FlamechartData>,
+  panAndZoomState: PanAndZoomState,
+  stackSectionBaseY: number,
+): ReactHoverContextInfo | null {
+  const {canvasMouseX, canvasMouseY, offsetY} = panAndZoomState;
+
+  const layerIndex = Math.floor(
+    (canvasMouseY + offsetY - stackSectionBaseY) / FLAMECHART_FRAME_HEIGHT,
+  );
+  const layer = flamechart.layers[layerIndex];
+
+  if (!layer) {
+    return null;
+  }
+
+  let startIndex = 0;
+  let stopIndex = layer.length - 1;
+  while (startIndex <= stopIndex) {
+    const currentIndex = Math.floor((startIndex + stopIndex) / 2);
+    const flamechartNode = layer[currentIndex];
+
+    const {end, start} = flamechartNode;
+
+    const width = durationToWidth((end - start) / 1000, panAndZoomState);
+    const x = Math.floor(timestampToPosition(start / 1000, panAndZoomState));
+
+    if (x <= canvasMouseX && x + width >= canvasMouseX) {
+      return {
+        event: null,
+        flamechartNode,
+        measure: null,
+        lane: null,
+        data,
+      };
+    }
+
+    if (x > canvasMouseX) {
+      stopIndex = currentIndex - 1;
+    } else {
+      startIndex = currentIndex + 1;
+    }
+  }
+
+  return null;
+}
+
+/**
+ * Returns a hover context object if the cursor is hovering over a React
+ * event/measure or a Flamechart node, otherwise returns null.
+ */
+export function getHoveredEvent(
+  schedulerCanvasHeight: number,
+  data: $ReadOnly<ReactProfilerDataV2>,
+  flamechart: $ReadOnly<FlamechartData>,
+  panAndZoomState: PanAndZoomState,
+): ReactHoverContextInfo | null {
+  const {canvasMouseY, offsetY} = panAndZoomState;
+
+  // These variables keep track of the current vertical stack sections' base and
+  // max Y coordinates. For example, if we're at the React event row, these are
+  // what the values represent:
+  // ┌-----------------------------------
+  // | t⁰    t¹    t²    ...
+  // ├---------------------------------- <- stackSectionBaseY
+  // |
+  // | <events...>
+  // |
+  // ├---------------------------------- <- stackSectionMaxY
+  // | <measures...>
+  // | <measures...>
+  // | <measures...>
+  // ├----------------------------------
+  // | <flame graph...>
+  // |
+  // |
+  // └----------------------------------
+  let stackSectionBaseY: number;
+  let stackSectionMaxY: number = 0;
+
+  // Header section: do nothing
+  stackSectionBaseY = stackSectionMaxY;
+  stackSectionMaxY += HEADER_HEIGHT_FIXED;
+  if (canvasMouseY < stackSectionMaxY) {
+    return null;
+  }
+
+  // ReactEvent row
+  stackSectionBaseY = stackSectionMaxY;
+  stackSectionMaxY += EVENT_ROW_HEIGHT_FIXED;
+  if (canvasMouseY + offsetY < stackSectionMaxY) {
+    return getHoveredReactEvent(data, panAndZoomState);
+  }
+
+  // ReactMeasure lanes
+  stackSectionBaseY = stackSectionMaxY;
+  stackSectionMaxY += schedulerCanvasHeight;
```

Comment:
The name `schedulerCanvasHeight` doesn't really make sense here. We'll fix this after #29 is done as this variable is also used in App.js and CanvasPage.js and is likely to be a deeper issue. I've opened #39 with more detail.